logger/logger.c

```diff
@@ -0,0 +1,267 @@
+		dma_log.header.timestamp,
+		entry.file_name,
+		entry.header.line_idx);
+	
+	switch (entry.header.params_num){
+	case 0:
+		fprintf(stdout, "%s", entry.text);
+		break;
+	case 1:
+		fprintf(stdout, entry.text, entry.params[0]);
+		break;
+	case 2:
+		fprintf(stdout, entry.text, entry.params[0],
+			entry.params[1]);
+		break;	
+	case 3:
+		fprintf(stdout, entry.text, entry.params[0], entry.params[1],
+			entry.params[2]);
+		break;
+	}
+	fprintf(stdout, "\n");
+}
+
+static int fetch_entry(FILE *f_ldc, FILE *f_dma, uint32_t base_address,
+	uint32_t data_offset, struct dma_log dma_log)
+{
+
+	struct ldc_entry entry;
+	long int padding;
+
+	uint32_t entry_offset;
+	uint32_t text_len;
+
+	int ret;
+
+	entry.file_name = NULL;
+	entry.text = NULL;
+	entry.params = NULL;
+	
+	/* evaluate entry offset in input file */
+	entry_offset = dma_log.address - base_address;
+	
+	/* set file position to beginning of processed entry */
+	fseek(f_ldc, entry_offset + data_offset, SEEK_SET);
+	
+	/* fetching elf header params */
+	ret = fread(&entry.header, sizeof(entry.header), 1, f_ldc);
+	if (!ret) {
+		ret = -EINVAL;
+		goto out;
+	}
+	
+	entry.file_name = (char *) malloc(entry.header.file_name_len);
+	if (!entry.file_name){
+		fprintf(stderr, "error: can't allocate %d byte for "
+			"entry.file_name\n", entry.header.file_name_len);
+		ret = -ENOMEM;
+		goto out;
+	}
+	
+	ret = fread(entry.file_name, sizeof(char), entry.header.file_name_len,
+		f_ldc);
+	if (ret != entry.header.file_name_len) {
+		ret = -EINVAL;
+		goto out;
+	}
+
+	/* padding - sequences of chars are aligned to DWORDS */
+	fseek(f_ldc, CEIL(entry.header.file_name_len, sizeof(uint32_t)) *
+		sizeof(uint32_t) - entry.header.file_name_len, SEEK_CUR);
+	
+	/* fetching text length */
+	ret = fread(&entry.text_len, sizeof(entry.text_len), 1, f_ldc);
+	if (!ret) {
+		ret = -EINVAL;
+		goto out;
+	}
+	
+	/* fetching text */
+	entry.text = (char *) malloc(entry.text_len);
+	if (entry.text == NULL) {
+		fprintf(stderr, "error: can't allocate %d byte for "
+			"entry.text\n", entry.text_len);
+		ret = -ENOMEM;
+		goto out;
+	}
+	
+	ret = fread(entry.text, sizeof(char), entry.text_len, f_ldc);
+	if (ret != entry.text_len) {
+		ret = -EINVAL;
+		goto out;
+	}
+	
+	/* fetching entry params from dma dump */
+	entry.params = (uint32_t *) malloc(sizeof(uint32_t) *
+		entry.header.params_num);
+	ret = fread(entry.params, sizeof(uint32_t), entry.header.params_num, 
+		f_dma);
+	if (ret != entry.header.params_num) {
+		ret = -EINVAL;
+		goto out;
+	}
+	
+	/* printing entry content */
+	print_entry_params(dma_log, entry);
+	
+	/* set f_ldc file position to the beginning */
+	rewind(f_ldc);
+	
+	ret = 0;
+out:
+	/* free alocated memory */
+	free(entry.params);
+	free(entry.text);
+	free(entry.file_name);
+
+	return ret;
+}
+
+int main(int argc, char *argv[])
+{
+	struct dma_log dma_log;
+	struct snd_sof_logs_header snd;
+
+	int ret;
+
+	const char *ldc_dir = argv[1];
+	const char *dma_dump = argv[2];
```

Comment:
Can you take rmbox as reference to have use getopt for the function.
Or just replace the old rmbox.

```c
	while ((opt = getopt(argc, argv, "ho:i:s:m:c:t")) != -1) {
		switch (opt) {
		case 'o':
			out_file = optarg;
			break;
		case 'i':
			in_file = optarg;
			break;
		case 't':
			trace = 1;
			break;
		case 'c':
			clk = atof(optarg);
			break;
		case 's':
			return snapshot(optarg);
		case 'h':
		default: /* '?' */
			usage(argv[0]);
		}
	}
```

Reply:
@bkokoszx yes pls use getopt, no need to replace rmbox atm as we need it for legacy. I want to merge this PR today.